### PR TITLE
Fix samples without patient middle name

### PR DESCRIPTION
**Pull request by xispa**
## Description of the issue/feature this PR addresses

This Pull Request fixes an *error while rendering senaite.core.sampleheader* in sample view when the sample does not have a middle name set.

## Current behavior before PR

An *error while rendering senaite.core.sampleheader* is displayed instead of a sample header

## Desired behavior after PR is merged

The sample header is displayed correctly

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: src/senaite/patient/content/fields.py [hunk 1]
```diff
@@ -75,11 +75,28 @@ class FullnameField(ExtensionField, ObjectField):
...
     security = ClassSecurityInfo()
 
-    def get(self, instance, **kwargs):
-        val = super(FullnameField, self).get(instance, **kwargs)
-        if isinstance(val, six.string_types):
-            val = {"firstname": val, "middlename": "", "lastname": ""}
-        return val
+    def set(self, instance, value, **kwargs):
+        keys = ["firstname", "middlename", "lastname"]
+        val = dict([(key, "") for key in keys])
```
Comment (ramonski) [inline src/senaite/patient/content/fields.py:80]: use `dict.fromkeys(["firstname", "middlename", "lastname"], "")` here

File: src/senaite/patient/content/fields.py [hunk 2]
```diff
@@ -75,11 +75,28 @@ class FullnameField(ExtensionField, ObjectField):
...
+
+        # Maybe fullname mode
+        if isinstance(value, six.string_types):
+            val.update({"firstname": value})
```
Comment (ramonski) [inline src/senaite/patient/content/fields.py:84]: use here `val["firstname"] = value`

File: src/senaite/patient/content/fields.py [hunk 3]
```diff
@@ -75,11 +75,28 @@ class FullnameField(ExtensionField, ObjectField):
...
+
+        # Ensure the stored object is from dict type. The received value can
+        # be from ZPublisher record type and since the field inherits from
+        # ObjectField, no conversion/decoding is done by the super field
+        value = dict(value) if value else {}
+
+        # Ensure the value contains expected keys
+        val.update(value)
```
Comment (ramonski) [inline src/senaite/patient/content/fields.py:92]: I would actually do something like this:
```python
try:
    for key, value in value.items():
        if key not in val:
            continue
        val[key] = value
except AttributeError:
    pass

File: src/senaite/patient/content/fields.py [hunk 4]
```diff
@@ -75,11 +75,28 @@ class FullnameField(ExtensionField, ObjectField):
...
+
+        # Set default if no values for any of the keys
+        values = filter(None, [val.get(key) for key in keys])
```
Comment (ramonski) [inline src/senaite/patient/content/fields.py:95]: Better use the built in `values()`:
```python
values = filter(None, val.values())
```

File: src/senaite/patient/upgrade/v01_04_000.py [hunk 5]
```diff
@@ -135,11 +138,76 @@ def fix_samples_middlename(tool):
...
+            # found some ZPublisher records!!
+            # (Pdb++) type(value)
+            # <class 'ZPublisher.HTTPRequest.record'>
+            try:
+                value = dict(value)
+            except:
+                str_type = repr(type(value))
+                logger.error("Type not supported: {}".format(str_type))
+                obj._p_deactivate()
+                continue
```
Comment (ramonski) [inline src/senaite/patient/upgrade/v01_04_000.py:195]: Maybe here better the explicit approach:
```python
keys = ["firstname", "middlename", "lastname"]
try:
    for k, v in value.items():
        if k not in keys:
            continue
        value[k] = v
except AttributeError:
    pass
```

**Review by ramonski — CHANGES_REQUESTED** [review]
Some changes are required, but I think these are worth it. Thanks for your efforts!

File: src/senaite/patient/upgrade/v01_04_000.py (hunk 5 above)
Comment (xispa) [reply src/senaite/patient/upgrade/v01_04_000.py:186]: It directly relies now on the field setter

**Review by ramonski — APPROVED** [review]

**Pull request merged**